pkg/collector/corechecks/containers/docker/check_metrics_extension.go

```diff
@@ -63,11 +64,50 @@ func (dn *dockerCustomMetricsExtension) Process(tags []string, container *worklo
 	}
 
 	if containerStats.CPU != nil {
-		dn.sender(dn.aggSender.Gauge, "docker.cpu.shares", containerStats.CPU.Shares, tags)
+		// Things to note about the "docker.cpu.shares" metric:
+		// - In cgroups v1 the metric that indicates how to share CPUs when
+		// there's contention is called shares. In v2 it's weight.
+		// - The idea is the same. The value of CPU shares and weight doesn't
+		// have any meaning by itself, it needs to be compared against the
+		// shares/weight of the other containers of the same host.
+		// - CPU shares and weight have different default values and different
+		// range of valid values. The range for shares is [2,262144], for weight
+		// it is [1,10000].
+		// - Even when using cgroups v2, the "docker run" command only accepts
+		// cpu shares as a parameter. "docker inspect" also shows shares. The
+		// formulas used to convert between shares and weights are these:
+		// https://github.com/kubernetes/kubernetes/blob/release-1.28/pkg/kubelet/cm/cgroup_manager_linux.go#L565
+		// - Because docker shows shares everywhere regardless of the cgroup
+		// version and "docker.cpu.shares" is a docker-specific metric, we think
+		// that it is less confusing to always report shares to match what
+		// the docker client reports.
+		// - "docker inspect" reports 0 shares when the container is created
+		// without specifying the number of shares. When that's the case, the
+		// default applies: 1024 for shares and 100 for weight.
+		// - The value emitted by the check is not exactly the same as in
+		// Docker because of the rounding applied in the conversions. Example:
+		//   - Run a container with 2048 shares in a system with cgroups v2.
+		//   - The 2048 shares are converted to weight in cgroups v2:
+		//     weight = (((shares - 2) * 9999) / 262142) + 1 = 79.04 (cgroups rounds to 79)
+		//   - This check converts the weight to shares again to report the same as in docker:
+		//     shares = (((weight - 1) * 262142) / 9999) + 2 = 2046.91 (will be rounded to 2047, instead of the original 2048).
+
+		var cpuShares float64
+		if containerStats.CPU.Shares != nil {
+			cpuShares = *containerStats.CPU.Shares
+		} else if containerStats.CPU.Weight != nil {
+			cpuShares = math.Round(cpuWeightToCPUShares(*containerStats.CPU.Weight))
+		}
+		dn.sender(dn.aggSender.Gauge, "docker.cpu.shares", &cpuShares, tags)
```

Comment:
There is a small behavior difference here compared to the previous code, but I’m not sure if it’s an issue.

With the previous code:
If
* `containerStats.CPU != nil` but
* `containerStats.CPU.Shares == nil`
no `docker.cpu.shares` metric was emitted because `dn.sender` sees that its 3rd argument is `nil`.

With the new code,
If
* `containerStats.CPU != nil` but
* `containerStats.CPU.Shares == nil` and `containerStats.CPU.Weight == nil`
then, we’ll emit a `docker.cpu.shares` metric with a value of `0` because `&cpuShares` isn’t `nil`.

Reply:
Thanks for pointing this out. I'm not sure if it's possible to have CPU != nil and both Shares and Weight nil, but I've changed this to have the same behavior as before, just in case.